File: Data Structures and Problem Solving Using Cpp 2nd/MkChnge.cpp

#ifdef USE_DOT_H
    #include <iostream.h>
#else
    #include <iostream>
    #if !defined( __BORLANDC__ ) || __BORLANDC__ >= 0x0530
        using namespace std;      // Borland 5.0 has a bug
    #endif
#endif
    
#ifndef SAFE_STL
    #include <vector>
    using std::vector;
#else
    #include "vector.h"
    #include "StartConv.h"
#endif

// Dynamic programming algorithm to solve change-making problem.
// As a result, the coinsUsed array is filled with the
// minimum number of coins needed for change from 0->maxChange
// and lastCoin contains one of the coins needed to make the change.
void makeChange( const vector<int> & coins, int maxChange,
                       vector<int> & coinsUsed, vector<int> & lastCoin )
{
    int differentCoins = coins.size( );
    coinsUsed.resize( maxChange + 1 );
    lastCoin.resize( maxChange + 1 );

    coinsUsed[ 0 ] = 0; lastCoin[ 0 ] = 1;

    for( int cents = 1; cents <= maxChange; cents++ )
    {
        int minCoins = cents, newCoin = 1;

        for( int j = 0; j < differentCoins; j++ )
        {
            if( coins[ j ] > cents )   // Can't use coin j
                continue;
            if( coinsUsed[ cents - coins[ j ] ] + 1 < minCoins )
            {
                minCoins = coinsUsed[ cents - coins[ j ] ] + 1;
                newCoin  = coins[ j ];
            }
        }

        coinsUsed[ cents ] = minCoins;
        lastCoin[ cents ]  = newCoin;
    }
}


// Simple test program
int main( )
{
        // The coins and the total amount of change
    vector<int> coins;
    coins.push_back( 1 );
    coins.push_back( 5 );
    coins.push_back( 10 );
    coins.push_back( 21 );
    coins.push_back( 25 );

    vector<int> used;
    vector<int> last;

    for( int change = 4; change < 100; change += 3 )
    {
        makeChange( coins, change, used, last );

        cout << "Total cents: " << change << "; Best is " << used[ change ]  << ": ";

        for( int i = change; i > 0; i -= last[ i ] )
            cout << last[ i ] << " ";
        cout << endl;
    }

    return 0;
}


#ifdef SAFE_STL
    #include "EndConv.h"
#endif

File: Data Structures and Problem Solving Using Java 4th/MkChnge.java
public final class MkChnge
{
    // Dynamic programming algorithm to solve change making problem.
    // As a result, the coinsUsed array is filled with the
    // minimum number of coins needed for change from 0 -> maxChange
    // and lastCoin contains one of the coins needed to make the change.

    public static void makeChange( int [ ] coins, int differentCoins,
                int maxChange, int [ ] coinsUsed, int [ ] lastCoin )
    {
        coinsUsed[ 0 ] = 0; lastCoin[ 0 ] = 1;

        for( int cents = 1; cents <= maxChange; cents++ )
        {
            int minCoins = cents;
            int newCoin  = 1;

            for( int j = 0; j < differentCoins; j++ )
            {
                if( coins[ j ] > cents )   // Cannot use coin j
                    continue;
                if( coinsUsed[ cents - coins[ j ] ] + 1 < minCoins )
                {
                    minCoins = coinsUsed[ cents - coins[ j ] ] + 1;
                    newCoin  = coins[ j ];
                }
            }

            coinsUsed[ cents ] = minCoins;
            lastCoin[ cents ]  = newCoin;
        }
    }

    // Simple test program
    public static void main( String [ ] args )
    {
        // The coins and the total amount of change
        int numCoins = 5;
        int [ ] coins = { 1, 5, 10, 21, 25 };
        int change = 0;

        if( args.length == 0 )
        {
            System.out.println( "Supply a monetary amount on the command line" );
            System.exit( 0 );
        }

        try
          { change = Integer.parseInt( args[ 0 ] ); }
        catch( NumberFormatException e )
        {
            System.out.println( e );
            System.exit( 0 ); 
        }

        int [ ] used = new int[ change + 1 ];
        int [ ] last = new int[ change + 1 ];

        makeChange( coins, numCoins, change, used, last );

        System.out.println( "Best is " + used[ change ] + " coins" );

        for( int i = change; i > 0; )
        {
            System.out.print( last[ i ] + " " );
            i -= last[ i ];
        }
        System.out.println( );
    }
}
